File: 0912 sort-an-array/solution.go
func sortArray(nums []int) []int {
    mergeSort(nums,0,len(nums)-1)
    return nums
}
func mergeSort(nums []int,left,right int) {
    if left>=right {
        return
    }
    mid:=left+(right-left)/2
    mergeSort(nums,left,mid)
    mergeSort(nums,mid+1,right)
    merge(nums,left,mid,right)
}
func merge(nums []int,left,mid,right int) {
    res:=make([]int,right-left+1)
    i,j,k:=left,mid+1,0
    for i<=mid && j<=right {
        if nums[i]<nums[j] {
            res[k]=nums[i]
            i++
        } else {
            res[k]=nums[j];
            j++
        }
        k++
    }
    for i<=mid {
        res[k]=nums[i]
        i++
        k++
    }
    for j<=right {
        res[k] = nums[j]
        k++
        j++
    }
    for t:=0;t<len(res);t++ {
        nums[left+t]=res[t]
    }
}

File: 0912 sort-an-array/solution.cs
public class Solution {
    public int[] SortArray(int[] nums) {
        // Array.Sort(nums);
        mergeSort(nums,0,nums.Length-1);
        return nums;
    }
    private void mergeSort(int[] nums,int left,int right) {
        if(left>=right) return;
        int mid=left+(right-left)/2;
        mergeSort(nums,left,mid);
        mergeSort(nums,mid+1,right);
        merge(nums,left,mid,right);
    }
    private void merge(int[] nums,int left,int mid,int right) {
        int i=left,j=mid+1,k=0;
        int[] res = new int[right-left+1];
        while(i<=mid && j<=right) {
            if(nums[i]<nums[j]) res[k++]=nums[i++]; 
            else res[k++]=nums[j++];
        }
        while(i<=mid) res[k++]=nums[i++];
        while(j<=right) res[k++]=nums[j++];
        for(i=0;i<res.Length;i++) nums[left+i]=res[i];
    }
}
